File: python/00019_remove_nth_node_from_end_of_list.py
#       Python
# Time complexity - O (N)
# Space complexity - O (1)

# example: 1 -> 2 -> 3 -> 4 -> 5; n = 2
# result:  1 -> 2 -> 3 -> 5; n = 2

# example: pointer (0) -> 1 -> 2 -> 3 -> 5; n = 2

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        pointer = ListNode(0)
        pointer.next = head
        fast = pointer
        slow = pointer

        # Move fast pointer n+1 steps ahead
        for _ in range(n + 1):
            fast = fast.next

        # Move both pointers until fast reaches the end
        while fast: # size - n
            fast = fast.next
            slow = slow.next

        # Remove the nth node from the end
        slow.next = slow.next.next

        return pointer.next

File: java/00019_remove_nth_node_from_end_of_list.java
//       Java
// Time complexity - O (N)
// Space complexity - O (1)

// example: 1 -> 2 -> 3 -> 4 -> 5; n = 2
// result:  1 -> 2 -> 3 -> 5; n = 2

//  example: pointer (0) -> 1 -> 2 -> 3 -> 5; n = 2

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        ListNode fast = dummy;
        ListNode slow = dummy;
        dummy.next = head;
        for (int i = 0; i < n + 1; i++) {
            fast = fast.next;
        }
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        slow.next = slow.next.next;
        return dummy.next;
    }
}